lib/srv/reexec.go

```diff
@@ -326,6 +332,28 @@ func RunCommand() (errw io.Writer, code int, err error) {
 		pamEnvironment = pamContext.Environment()
 	}
 
+	// Set SELinux context for the child process if SELinux support is
+	// enabled so the child process will be running with the correct SELinux
+	// user, role and domain.
+	if c.SetSELinuxContext {
+		seContext, err := selinux.UserContext(c.Login)
+		if err != nil {
+			return errorWriter, teleport.RemoteCommandFailure, trace.Wrap(err, "failed to get SELinux context of login user")
+		}
+
+		// SetExecLabel changes the SELinux exec context for the
+		// calling thread only, so we need to ensure that is the
+		// thread that will create the child. We don't ever unlock
+		// the thread as we're exiting after the child exits, and
+		// we want to avoid another goroutine getting denied due to
+		// running on this thread with a different (likely much more
+		// restrictive)SELinux context.
+		runtime.LockOSThread()
```

Comment:
I think I'd rather see a justification for why locking the thread is the caller's responsibility like for PAM, otherwise locking the thread here is meaningless, the function will return without unlocking the OS thread but the goroutine is still usable by the caller.

Perhaps this should be part of `RunCommand`'s docs instead? Something like "This function might change system state related to the process and/or the thread for PAM and SELinux so it should only be called from the main goroutine locked as per runtime.LockOSThread".

Reply:
As you mentioned in our slack convo, we only have LockOSThread in init in pam builds, so we should leave the LockOSThread here _and_ add the comment.